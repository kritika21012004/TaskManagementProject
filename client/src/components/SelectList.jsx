import { Listbox, Transition } from "@headlessui/react";
import React, { Fragment } from "react";
import { BsChevronExpand } from "react-icons/bs";
import { MdCheck } from "react-icons/md";
import "../styles/SelectList.css"

const SelectList = ({ lists, selected, setSelected, label }) => {
  return (
    <div className='w-full'>
      {label && <p className='label'>{label}</p>}

      <Listbox value={selected} onChange={setSelected}>
        <div className='listbox-outer-div'>
          <Listbox.Button className='listbox-button'>
            <span className='block-truncate'>{selected}</span>
            <span className='selectlist-expand-span'>
              <BsChevronExpand
                className='selectlist-expand'
                aria-hidden='true'
              />
            </span>
          </Listbox.Button>
          <Transition
            as={Fragment}
            leave='leave'
            leaveFrom='leavefrom'
            leaveTo='leaveto'
          >
            <Listbox.Options className='list-map'>
              {lists.map((list, index) => (
                <Listbox.Option
                  key={index}
                  className={({ active }) =>
                    `listmap ${
                      active ? "color-class" : "text-gray-900"
                    }`
                  }
                  value={list}
                >
                  {({ selected }) => (
                    <>
                      <span
                        className={`block-truncate ${
                          selected ? "font-medium" : "font-normal"
                        }`}
                      >
                        {list}
                      </span>
                      {selected ? (
                        <span className='mdcheck-span'>
                          <MdCheck className='h-5-w-5' aria-hidden='true' />
                        </span>
                      ) : null}
                    </>
                  )}
                </Listbox.Option>
              ))}
            </Listbox.Options>
          </Transition>
        </div>
      </Listbox>
    </div>
  );
};

export default SelectList;