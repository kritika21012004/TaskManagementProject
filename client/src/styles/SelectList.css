.font-medium {
    font-weight: 500;
}

.font-normal {
    font-weight: 400;
}

.block-truncate {
    display: block;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.mdcheck-span {
    display: flex;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    padding-left: 0.75rem;
    align-items: center;
}

.h-5-w-5 {
    width: 1.25rem;
    height: 1.25rem;
}
.listmap{
    position: relative; 
padding-top: 0.5rem;
padding-bottom: 0.5rem; 
padding-right: 1rem; 
padding-left: 2.5rem; 
cursor: default; 
user-select: none; 
}
.w-full{
    width:100%;
}
.label{
    color: #6B7280; 
    padding:0.5rem;
}
.listbox-outer-div{
    position: relative; 
margin-top: 0.25rem; 
}
.listbox-button{
    position: relative; 
padding-top: 0.625rem;
padding-bottom: 0.625rem; 
padding-left: 0.75rem;
padding-right: 0.75rem; 
padding-left: 0.75rem; 
padding-right: 2.5rem; 
border-radius: 0.25rem; 
border-width: 1px; 
border-color: #D1D5DB; 
width: 100%; 
text-align: left; 
background-color: #ffffff; 
cursor: default; 

@media (min-width: 640px) { 
  font-size: 0.875rem;
line-height: 1.25rem; 
 }

@media (min-width: 1536px) { 
   }

}

.selectlist-expand-span{
    display: flex; 
position: absolute; 
top: 0;
bottom: 0; 
right: 0; 
padding-right: 0.5rem; 
align-items: center; 
pointer-events: none; 
}
.selectlist-expand{
width: 1.25rem; 
height: 1.25rem; 
color: #9CA3AF; 
}
.leave{
    transition-property: background-color, border-color, color, fill, stroke, opacity, box-shadow, transform;
transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
transition-duration: 300ms; 
transition-duration: 100ms; 
transition-timing-function: cubic-bezier(0.4, 0, 1, 1); 
}
.leavefrom{
    opacity: 1; 
}
.leaveto{
    opacity: 0; 
}
.list-map{
    overflow: auto; 
position: absolute; 
z-index: 50; 
padding-top: 0.25rem;
padding-bottom: 0.25rem; 
margin-top: 0.25rem; 
border-radius: 0.375rem; 
box-shadow: var(--tw-ring-inset) 0 0 0 calc(1px + var(--tw-ring-offset-width)) var(--tw-ring-color); 
width: 100%; 
font-size: 1rem;
line-height: 1.5rem; 
background-color: #ffffff; 
box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05); 

@media (min-width: 640px) { 
  font-size: 0.875rem;
line-height: 1.25rem; 
 }

}
.text-gray-900{
    color: #111827; 
}
.leaveto{
    opacity: 0; 
}
.color-class{
    background-color: #FFECB3;
    color: #FF5722;
}







