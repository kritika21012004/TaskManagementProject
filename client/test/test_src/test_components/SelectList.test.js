import React from 'react';
import SelectList from '../../../src/components/SelectList';
import renderer from 'react-test-renderer';

describe('SelectList', () => {
  const lists = ['Option 1', 'Option 2', 'Option 3'];
  const label = 'Test Label';

  it('renders correctly with initial props', () => {
    const setSelected = jest.fn();

    const tree = renderer
      .create(<SelectList lists={lists} selected={lists[0]} setSelected={setSelected} label={label} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("handles no props gracefully", () => {
    const defaultProps = {
      setSelected: jest.fn(),
      label: "Default Label",
      lists: [],
    };
  
    expect(() => renderer.create(<SelectList { ...defaultProps }/>)).not.toThrow();
  });
  
  it("handles non-matching selected prop gracefully", () => {
    const setSelected = jest.fn();
    const lists = ["Option 1", "Option 2", "Option 3"];
    const label = "Test Label";
    const selected = "Option 4"; // This option doesn't exist in lists
    expect(() => renderer.create(
      <SelectList
        lists={lists}
        selected={selected}
        setSelected={setSelected}
        label={label}
      />
    )).not.toThrow();
  });

  
  it("handles missing lists prop gracefully", () => {
    const setSelected = jest.fn();
    const label = "Test Label";
    const lists = [];
    expect(() =>
    renderer.create(
    <SelectList setSelected={setSelected} label={label} lists={lists}/>
    )
    ).not.toThrow();
    });
  
  it("handles missing label prop gracefully", () => {
    const setSelected = jest.fn();
    const lists = ["Option 1", "Option 2", "Option 3"];
    expect(() => renderer.create(<SelectList setSelected={setSelected} lists={lists}/>)).not.toThrow();
  });
  it("updates correctly when new props are passed", () => {
    const setSelected = jest.fn();
    const label = "Test Label";
    const newLabel = "New Test Label";
    const lists = ['test1', 'test2'];
  
    const component = renderer.create(
      <SelectList
        lists={lists}
        selected={lists[0]}
        setSelected={setSelected}
        label={label}
      />
    );
  
    component.update(
      <SelectList
        lists={lists}
        selected={lists[1]}
        setSelected={setSelected}
        label={newLabel}
      />
    );
  
    const instance = component.root;
    expect(instance.props.selected).toBe(lists[1]);
    expect(instance.props.label).toBe(newLabel);
  });

});
