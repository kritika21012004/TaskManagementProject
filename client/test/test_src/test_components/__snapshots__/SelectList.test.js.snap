// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectList renders correctly with initial props 1`] = `
<div
  className="w-full"
>
  <p
    className="label"
  >
    Test Label
  </p>
  <div
    className="listbox-outer-div"
    data-headlessui-state=""
  >
    <button
      aria-expanded={false}
      aria-haspopup="listbox"
      autoFocus={false}
      className="listbox-button"
      data-headlessui-state=""
      id="headlessui-listbox-button-:r2:"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onPointerDown={[Function]}
      onPointerUp={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <span
        className="block-truncate"
      >
        Option 1
      </span>
      <span
        className="selectlist-expand-span"
      >
        <svg
          aria-hidden="true"
          className="selectlist-expand"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.646 9.146a.5.5 0 0 1 .708 0L8 12.793l3.646-3.647a.5.5 0 0 1 .708.708l-4 4a.5.5 0 0 1-.708 0l-4-4a.5.5 0 0 1 0-.708m0-2.292a.5.5 0 0 0 .708 0L8 3.207l3.646 3.647a.5.5 0 0 0 .708-.708l-4-4a.5.5 0 0 0-.708 0l-4 4a.5.5 0 0 0 0 .708"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`SelectList renders correctly with initial props 2`] = `
<div
  className="w-full"
>
  <p
    className="label"
  >
    Test Label
  </p>
  <div
    className="listbox-outer-div"
    data-headlessui-state=""
  >
    <button
      aria-expanded={false}
      aria-haspopup="listbox"
      autoFocus={false}
      className="listbox-button"
      data-headlessui-state=""
      id="headlessui-listbox-button-:r5:"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onPointerDown={[Function]}
      onPointerUp={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <span
        className="block-truncate"
      >
        Option 1
      </span>
      <span
        className="selectlist-expand-span"
      >
        <svg
          aria-hidden="true"
          className="selectlist-expand"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.646 9.146a.5.5 0 0 1 .708 0L8 12.793l3.646-3.647a.5.5 0 0 1 .708.708l-4 4a.5.5 0 0 1-.708 0l-4-4a.5.5 0 0 1 0-.708m0-2.292a.5.5 0 0 0 .708 0L8 3.207l3.646 3.647a.5.5 0 0 0 .708-.708l-4-4a.5.5 0 0 0-.708 0l-4 4a.5.5 0 0 0 0 .708"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
